### PR TITLE
[feature] EvidenceItem and Thermometer

### LINKED ISSUE #145: Implement Base EvidenceItem class
- inherits item
- 2 states for each implemented EvidenceItem
- Use method will not be implemented in exact functionality atm, at least has the ability to change between states

### LINKED ISSUE #146: Implement Evidence Item temperature

### PR DESCRIPTION
**Pull request by JahnsMichael**
### Main changes

- add EvidenceItem base class
- add new Ghost Interaction Simulation Input Action
- add Thermometer temporary implementation

### Tasks List

- [close #145]
- [close #146]

### Attachments

https://user-images.githubusercontent.com/60375650/152746293-4f34c0c2-09e1-4fb5-b55f-9b853c9a6dda.mp4

Implemented State Transition
- Use --> Active (Yellow)
- When Active then Interacted by Ghost --> Evidence A (Blue)
- When Evidence A then Interacted by Ghost --> Evidence B (Red) (temporary/dummy behavior)
- Discard --> Inactive (Orange)
- When inactive then Interacted by Ghost --> Nothing

### CODE REVIEW

**Review by naufalhilmi12345 — CHANGES_REQUESTED** [review]
Changing adding newline at EvidenceItem.cs

File: Assets/Astralization/Scripts/Interactables/Item/EvidenceItem/EvidenceItem.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public void OnGhostInteraction()
+    {
+        if (state != EvidenceItemState.BASE) DetermineEvidence();
+    }
+
+    public abstract void DetermineEvidence();
+
+    public abstract void HandleChange();
+
+}
```
[the remark this answers is not in the document]
Comment (JahnsMichael) [reply Assets/Astralization/Scripts/Interactables/Item/EvidenceItem/EvidenceItem.cs]: done

**Review by naufalhilmi12345 — APPROVED** [review]
Cool! LGTM

File: Assets/Astralization/Scripts/Interactables/Item/EvidenceItem/Thermometer.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private void SetStateMaterial(Material stateMaterial)
+    {
+        MeshRenderer mesh = model.GetComponentInChildren<MeshRenderer>(true);
+        mesh.material = stateMaterial;
+        Debug.Log("[MAT] " + mesh.material.name);
+    }
+
+    public override void HandleChange()
+    {
+        switch (this.state)
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/Interactables/Item/EvidenceItem/Thermometer.cs:41]: can be refactored: make Dictionary<EvidenceItemState, Material> and use the dictionary to set state material

File: Assets/Astralization/Scripts/Interactables/Item/Item.cs [hunk 3]
```diff
@@ -60,4 +67,9 @@ public Sprite GetItemLogo()
     {
         return _itemLogo;
     }
+
+    public bool GetDiscardWhenUsed()
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/Interactables/Item/Item.cs:71]: rename to IsDiscardedWhenUsed. kalo returnnya boolean, lebih baik pake prefix Is/Are

File: Assets/Astralization/Scripts/Player/Inventory.cs [hunk 4]
```diff
@@ -199,7 +199,27 @@ public void UseActiveItem(InputAction.CallbackContext ctx)
         if (ctx.performed)
         {
             _activeItem?.Use();
+
             if (!_activeItem) Debug.Log("[ITEM] Missing active item");
+            else if (_activeItem.GetDiscardWhenUsed())
+            {
+                // Debug.Log("[INVENTORY] Discard " + _activeItem.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Scripts/Player/Inventory.cs:206]: refactor: extract line 141-155 to another function and call that function

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        yield return new WaitWhile(() => sceneLoaded == false);
+        GameObject thermometer = GameObject.Find("Thermometer");
+        float moveDuration = GetMovementDurationTowards(thermometer.transform);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.UseItem);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Active (Instance)", stateMaterial.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:71]: assert the item state

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        yield return new WaitWhile(() => sceneLoaded == false);
+        GameObject thermometer = GameObject.Find("Thermometer");
+        float moveDuration = GetMovementDurationTowards(thermometer.transform);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.DiscardItem);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Base (Instance)", stateMaterial.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:85]: assert the item state

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.UseItem);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Active (Instance)", stateMaterial.name);
+    }    
+    
+    [UnityTest]
+    public IEnumerator EvidenceItem_NotActivatedWhenUsed()
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:75]: rename to EvidenceItem_NotActivatedWhenDiscarded

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        yield return new WaitWhile(() => sceneLoaded == false);
+        GameObject thermometer = GameObject.Find("Thermometer");
+        float moveDuration = GetMovementDurationTowards(thermometer.transform);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.DiscardItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.SimulateGhostInteract);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Base (Instance)", stateMaterial.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:104]: assert the item state

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        GameObject thermometer = GameObject.Find("Thermometer");
+        float moveDuration = GetMovementDurationTowards(thermometer.transform);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.UseItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.SimulateGhostInteract);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.SimulateGhostInteract);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Positive (Instance)", stateMaterial.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:120]: assert the item state

File: Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        yield return new WaitWhile(() => sceneLoaded == false);
+        GameObject thermometer = GameObject.Find("Thermometer");
+        float moveDuration = GetMovementDurationTowards(thermometer.transform);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.MoveBack, false, moveDuration);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.PickItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.UseItem);
+        yield return SimulateInput(KeyboardMouseTestFixture.RegisteredInput.SimulateGhostInteract);
+        GameObject thermometerModel = GameObject.Find("OverworldItems/Thermometer/Model");
+        Material stateMaterial = thermometerModel.GetComponent<MeshRenderer>().material;
+        Assert.AreEqual("MAT_Thermometer_Negative (Instance)", stateMaterial.name);
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/EvidenceItemTest.cs:135]: assert the item state

**Review by rafiftaris — CHANGES_REQUESTED** [review]
dont forget the feature docs and add testcases to TC list 😉

**Comment by JahnsMichael** [comment]
> dont forget the feature docs and add testcases to TC list wink

done

**Review by rafiftaris — APPROVED** [review]

**Review by raihansyahdean — CHANGES_REQUESTED** [review]

File: Assets/Astralization/Scripts/Player/Inventory.cs [hunk 11]
```diff
@@ -132,28 +132,30 @@ public void PickItem(Item item)
         }
     }
 
-    public void DiscardItem(InputAction.CallbackContext ctx)
+    private void _DiscardItem()
```
[the remark this answers is not in the document]
Comment (JahnsMichael) [reply Assets/Astralization/Scripts/Player/Inventory.cs]: done

**Review by raihansyahdean — APPROVED** [review]
Nice

**Pull request merged**